Add documentation of component.utils exportFunction and cloneInto (#30877)
mdn/content: files/en-us/mozilla/add-ons/webextensions/content_scripts/cloneinto/index.md

---
title: cloneInto()
slug: Mozilla/Add-ons/WebExtensions/Content_scripts/cloneInto
page-type: webextension-api-function
browser-compat: webextensions.api.contentScriptGlobalScope.cloneInto
---

{{AddonSidebar()}}

This function provides a safe way to take an object defined in a privileged scope and create a [structured clone](/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm) of it in a less-privileged scope. It returns a reference to the clone:

```js
var clonedObject = cloneInto(myObject, targetWindow);
```

You can then assign the clone to an object in the target scope as an expando property, and scripts running in that scope can access it:

```js
targetWindow.foo = clonedObject;
```

This enables privileged code, such as an extension, to share an object with less-privileged code, such as a web page script.

## Syntax

```js-nolint
let clonedObject = cloneInto(
  obj,               // object
  targetScope,       // object

>   options            // object

Reviewer: ```suggestion
  options            // optional object
```

---

> title: components.utils.cloneInto()

Reviewer: ```suggestion
title: cloneInto()
```

slug: Mozilla/Add-ons/WebExtensions/Content_scripts/cloneInto
page-type: webextension-api-function

> browser-compat: webextensions.api.components.utils.cloneInto

Reviewer: ```suggestion
browser-compat: webextensions.api.ContentScriptGlobalScope.cloneInto
```
`components.utils` is an internal name and should not be exposed because it cannot be used. Here is an alternative suggestion, with a new made up name, ContentScriptGlobalScope, to have a namespace that can be referenced from mdn/content.

---

{{AddonSidebar()}}

This function provides a safe way to take an object defined in a privileged scope and create a [structured clone](/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm) of it in a less-privileged scope. It returns a reference to the clone:

```js
var clonedObject = cloneInto(myObject, targetWindow);
```

You can then assign the clone to an object in the target scope as an expando property, and scripts running in that scope can access it:

```js
targetWindow.foo = clonedObject;
```

This enables privileged code, such as an extension, to share an object with less-privileged code, such as a web page script.

## Syntax

```js-nolint
let clonedObject = Components.utils.cloneInto(
  obj,               // object
  targetScope,       // object
  options            // object
);
```

### Parameters

- `obj`
  - : `object`. The object to clone.
- `targetScope`
  - : `object`. The object to attach the object to.
- `options` {{optional_inline}}
  - : `object`. Options for the function.
    - `cloneFunctions` {{optional_inline}}

>       - : `Boolean`. Whether the object's functions should be cloned. Default to `false`. Cloned functions have the same semantics as functions exported using [`Components.utils.exportFunction`](/en-US/docs/Mozilla/Add-ons/WebExtensions/API/components/utils/exportFunction). See [Cloning objects that have functions](#Cloning_objects_that_have_functions). {{optional_inline}}

Reviewer: ```suggestion
      - : `boolean`. Whether the object's functions should be cloned. Default to `false`. Cloned functions have the same semantics as functions exported using [`exportFunction`](/en-US/docs/Mozilla/Add-ons/WebExtensions/API/Content_scripts/exportFunction). See [Cloning objects that have functions](#Cloning_objects_that_have_functions). {{optional_inline}}
```

    - `wrapReflectors` {{optional_inline}}

>       - : `boolean`. Whether objects reflected from C++, such as DOM objects, should be cloned. Defaults to `false`. See [Cloning objects that contain DOM elements](#Cloning_objects_that_contain_DOM_elements).

Reviewer: ```suggestion
      - : `boolean`. Whether DOM objects should be passed by reference instead of cloned. DOM objects are usually not clonable. Defaults to `false`. See [Cloning objects that contain DOM elements](#Cloning_objects_that_contain_DOM_elements).
```

### Return Value

A reference to the cloned object.

## Examples

> This add-on script creates an object, clones it into the content window and makes it a property of the content window global:

Reviewer: Can you replace all references of "add-on script" with "content script"? The original Components.utils.cloneInto method and its family of functions was documented in a general way because privileged code in add-ons, in general, could use this API anywhere. But in WebExtensions, only content scripts (and userScripts) can use it.

```js
// add-on script
var addonScriptObject = { greeting: "hello from add-on" };

> contentWindow.addonScriptObject = cloneInto(addonScriptObject, contentWindow);

Reviewer: ```suggestion
window.addonScriptObject = cloneInto(addonScriptObject, window);
```

Replace contentWindow with window everywhere in this article.

```

Scripts running in the page can access the object:

```js
// page script
button.addEventListener(
  "click",
  function () {
    console.log(window.addonScriptObject.greeting); // "hello from add-on"
  },
  false,
);
```

Of course, you don't have to assign the clone to the window itself; you can assign it to some other object in the target scope:

```js

> contentWindow.foo.addonScriptObject = cloneInto(

Reviewer: ```suggestion
// Content script
contentWindow.foo.addonScriptObject = cloneInto(
```

For clarity emphasize in this whole example whether a script is a content script or a page script.
